File: java/client/src/test/java/org/apache/rocketmq/client/java/metrics/ClientMeterTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.rocketmq.client.java.metrics;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import apache.rocketmq.v2.Endpoints;
import io.opentelemetry.api.metrics.Meter;
import io.opentelemetry.sdk.OpenTelemetrySdk;
import io.opentelemetry.sdk.metrics.SdkMeterProvider;
import io.opentelemetry.sdk.resources.Resource;
import org.apache.rocketmq.client.java.misc.ClientId;
import org.apache.rocketmq.client.java.tool.TestBase;
import org.junit.Test;

public class ClientMeterTest extends TestBase {

    @Test
    public void testShutdownWithEnabledMeter() {
        final SdkMeterProvider provider = SdkMeterProvider.builder().setResource(Resource.empty()).build();
        final OpenTelemetrySdk openTelemetry = OpenTelemetrySdk.builder().setMeterProvider(provider).build();
        Meter meter = openTelemetry.getMeter("test");
        final ClientId clientId = new ClientId();
        final ClientMeter clientMeter = new ClientMeter(meter, fakeEndpoints(), provider, clientId);
        assertTrue(clientMeter.isEnabled());
        clientMeter.shutdown();
    }

    @Test
    public void testShutdownWithDisabledMeter() {
        final ClientId clientId = new ClientId();
        final ClientMeter clientMeter = ClientMeter.disabledInstance(clientId);
        assertFalse(clientMeter.isEnabled());
        clientMeter.shutdown();
    }

    @Test
    public void testSatisfy() {
        final ClientId clientId = new ClientId();
        ClientMeter clientMeter = ClientMeter.disabledInstance(clientId);
        Metric metric = new Metric(apache.rocketmq.v2.Metric.newBuilder().setOn(false).build());
        assertTrue(clientMeter.satisfy(metric));

        metric = new Metric(apache.rocketmq.v2.Metric.newBuilder().setOn(true).build());
        assertTrue(clientMeter.satisfy(metric));

        final Endpoints pbEndpoints0 = fakePbEndpoints0();

        metric = new Metric(apache.rocketmq.v2.Metric.newBuilder().setOn(false).setEndpoints(pbEndpoints0).build());
        assertTrue(clientMeter.satisfy(metric));

        metric = new Metric(apache.rocketmq.v2.Metric.newBuilder().setOn(true).setEndpoints(pbEndpoints0).build());
        assertFalse(clientMeter.satisfy(metric));

        final org.apache.rocketmq.client.java.route.Endpoints endpoints =
            new org.apache.rocketmq.client.java.route.Endpoints(pbEndpoints0);

        final SdkMeterProvider provider = SdkMeterProvider.builder().setResource(Resource.empty()).build();
        final OpenTelemetrySdk openTelemetry = OpenTelemetrySdk.builder().setMeterProvider(provider).build();
        Meter meter = openTelemetry.getMeter("test");
        clientMeter = new ClientMeter(meter, endpoints, provider, clientId);

        metric = new Metric(apache.rocketmq.v2.Metric.newBuilder().setOn(false).build());
        assertFalse(clientMeter.satisfy(metric));

        metric = new Metric(apache.rocketmq.v2.Metric.newBuilder().setOn(true).build());
        assertFalse(clientMeter.satisfy(metric));

        metric = new Metric(apache.rocketmq.v2.Metric.newBuilder().setOn(false).setEndpoints(pbEndpoints0).build());
        assertFalse(clientMeter.satisfy(metric));

        metric = new Metric(apache.rocketmq.v2.Metric.newBuilder().setOn(true).setEndpoints(pbEndpoints0).build());
        assertTrue(clientMeter.satisfy(metric));

        final Endpoints pbEndpoints1 = fakePbEndpoints1();
        metric = new Metric(apache.rocketmq.v2.Metric.newBuilder().setOn(true).setEndpoints(pbEndpoints1).build());
        assertFalse(clientMeter.satisfy(metric));
    }
}

File: csharp/tests/ClientMeterTest.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using Microsoft.VisualStudio.TestTools.UnitTesting;
using OpenTelemetry;
using OpenTelemetry.Metrics;
using OpenTelemetry.Resources;
using Org.Apache.Rocketmq;
using Metric = Org.Apache.Rocketmq.Metric;
using Proto = Apache.Rocketmq.V2;

namespace tests
{
    [TestClass]
    public class ClientMeterTest
    {
        private MeterProvider CreateMeterProvider()
        {
            return Sdk.CreateMeterProviderBuilder()
                      .SetResourceBuilder(ResourceBuilder.CreateEmpty())
                      .Build();
        }

        [TestMethod]
        public void TestShutdownWithEnabledMeter()
        {
            var endpoints = new Endpoints(new Proto.Endpoints
            {
                Scheme = Proto.AddressScheme.Ipv4,
                Addresses = { new Proto.Address { Host = "127.0.0.1", Port = 8080 } }
            });
            var provider = CreateMeterProvider();
            var clientId = "testClientId";
            var clientMeter = new ClientMeter(endpoints, provider, clientId);
            Assert.IsTrue(clientMeter.Enabled);
            clientMeter.Shutdown();
        }

        [TestMethod]
        public void TestShutdownWithDisabledMeter()
        {
            var clientId = "testClientId";
            var clientMeter = ClientMeter.DisabledInstance(clientId);
            Assert.IsFalse(clientMeter.Enabled);
            clientMeter.Shutdown();
        }

        [TestMethod]
        public void TestSatisfy()
        {
            var clientId = "testClientId";
            var clientMeter = ClientMeter.DisabledInstance(clientId);

            var metric = new Metric(new Proto.Metric { On = false });
            Assert.IsTrue(clientMeter.Satisfy(metric));

            metric = new Metric(new Proto.Metric { On = true });
            Assert.IsTrue(clientMeter.Satisfy(metric));

            var endpoints0 = new Proto.Endpoints
            {
                Scheme = Proto.AddressScheme.Ipv4,
                Addresses = { new Proto.Address { Host = "127.0.0.1", Port = 8080 } }
            };

            metric = new Metric(new Proto.Metric { On = false, Endpoints = endpoints0 });
            Assert.IsTrue(clientMeter.Satisfy(metric));

            metric = new Metric(new Proto.Metric { On = true, Endpoints = endpoints0 });
            Assert.IsFalse(clientMeter.Satisfy(metric));

            var endpoints = new Endpoints(endpoints0);
            var provider = CreateMeterProvider();
            clientMeter = new ClientMeter(endpoints, provider, clientId);

            metric = new Metric(new Proto.Metric { On = false });
            Assert.IsFalse(clientMeter.Satisfy(metric));

            metric = new Metric(new Proto.Metric { On = true });
            Assert.IsFalse(clientMeter.Satisfy(metric));

            metric = new Metric(new Proto.Metric { On = false, Endpoints = endpoints0 });
            Assert.IsFalse(clientMeter.Satisfy(metric));

            metric = new Metric(new Proto.Metric { On = true, Endpoints = endpoints0 });
            Assert.IsTrue(clientMeter.Satisfy(metric));

            var endpoints1 = new Proto.Endpoints
            {
                Scheme = Proto.AddressScheme.Ipv4,
                Addresses = { new Proto.Address { Host = "127.0.0.2", Port = 8081 } }
            };
            metric = new Metric(new Proto.Metric { On = true, Endpoints = endpoints1 });
            Assert.IsFalse(clientMeter.Satisfy(metric));
        }
    }
}